CMakeLists.txt

```diff
@@ -12,6 +12,7 @@ SET(CMAKE_C_FLAGS "-std=c99 -Werror=implicit-function-declaration ${CMAKE_C_FLAG
 ENDIF (NOT WIN32)
 IF(CUDA_HAS_FP16 OR NOT ${CUDA_VERSION} LESS 7.5)
   SET(CMAKE_C_FLAGS "-DCUDA_HAS_FP16=1 ${CMAKE_C_FLAGS}")
+  SET(CMAKE_C_FLAGS "-DREDUCED_TYPE_SET=1 ${CMAKE_C_FLAGS}")
```

Comment:
Responding to questions:
1. "What's special about char, short, int, double to be considered for non-compilation, compared to e.g. byte?"
The goal was to provide a smaller size libraries for Android apps that use float tensors only.
 Code bloat from nvcc is huge. I have tried 'byte' and 'long' initially, too, but they were used internally by cutorch helper routines.

2. "The need to have #ifdef CUDA_HALF_TENSOR and #ifndef CUDA_NO_HALF for half types to be defined is going to have constant maintenance problems -- we already break cutorch regularly because of lack of CUDA_HALF_TENSOR..."
Actually,  hopefully CUDA_HALF_TENSOR could be retired soon - probably as soon as we have CPU Half integrated, so I don't see much issue here.
3. "Similar to 1) above, I don't see how people are going to remember that half (sort of), char, short, int, double are special and need to be guarded and other types don't."
I agree, those changes are not pretty, and require some re-thinking and/or refactoring beforehand. Part of the reason the changes are so uneven and scattered is because the code may be too expanded and some clever use of macros may help. 
I am going to withdraw this PR for now.